Xamarin.Forms.Platform.Android/Renderers/ListViewAdapter.cs

```diff
@@ -309,6 +311,68 @@ public override AView GetView(int position, AView convertView, ViewGroup parent)
 			return layout;
 		}
 
+		internal void InvalidatePrototypicalCellCache()
+		{
+			_prototypicalCellByTypeOrDataTemplate.Clear();
+		}
+
+		internal ITemplatedItemsList<Cell> GetTemplatedItemsListForPath(int indexPath)
+		{
+			var templatedItems = TemplatedItemsView.TemplatedItems;
+			if (_listView.IsGroupingEnabled)
+				templatedItems = (ITemplatedItemsList<Cell>)((IList)templatedItems)[indexPath];
+
+			return templatedItems;
+		}
+
+		internal DataTemplate GetDataTemplateForPath(int indexPath)
+		{
+			var templatedItemsList = GetTemplatedItemsListForPath(indexPath);
+			var item = templatedItemsList.ListProxy[indexPath];
+			return templatedItemsList.SelectDataTemplate(item);
+		}
+
+		internal Type GetItemTypeForPath(int indexPath)
+		{
+			var templatedItemsList = GetTemplatedItemsListForPath(indexPath);
+			var item = templatedItemsList.ListProxy[indexPath];
+			return item.GetType();
+		}
+
+		internal Cell GetCellForPath(int indexPath)
+		{
+			var templatedItemsList = GetTemplatedItemsListForPath(indexPath);
+			var cell = templatedItemsList[indexPath];
+			return cell;
+		}
+
+		internal Cell GetPrototypicalCell(int indexPath)
+		{
+			var itemTypeOrDataTemplate = default(object);
+
+			var cachingStrategy = _listView.CachingStrategy;
+			if (cachingStrategy == ListViewCachingStrategy.RecycleElement)
+				itemTypeOrDataTemplate = GetDataTemplateForPath(indexPath);
+
+			else if (cachingStrategy == ListViewCachingStrategy.RecycleElementAndDataTemplate)
+				itemTypeOrDataTemplate = GetItemTypeForPath(indexPath);
+
+			else // ListViewCachingStrategy.RetainElement
+				return GetCellForPosition(indexPath);
+
+			Cell protoCell;
+			if (!_prototypicalCellByTypeOrDataTemplate.TryGetValue(itemTypeOrDataTemplate, out protoCell))
+			{
+				// cache prototypical cell by item type; Items of the same Type share
+				// the same DataTemplate (this is enforced by RecycleElementAndDataTemplate)
+				protoCell = GetCellForPosition(indexPath);
+				_prototypicalCellByTypeOrDataTemplate[itemTypeOrDataTemplate] = protoCell;
+			}
+
+			var templatedItems = GetTemplatedItemsListForPath(indexPath);
+			return templatedItems.UpdateContent(protoCell, indexPath);
+		}
+
```

Comment:
Will rename `indexPath` to `index` and `ForPath` to `ForIndex` before merging. For now, leave as is so as not to trigger another test run.